crates/red_knot_python_semantic/resources/mdtest/narrow/type_guards.md

```diff
@@ -0,0 +1,227 @@
+# User-defined type guards
+
+User-defined type guards are functions of which the return type is either `TypeGuard[...]` or
+`TypeIs[...]`.
+
+## Display
+
+```py
+from knot_extensions import Intersection, Not, TypeOf
+from typing_extensions import TypeGuard, TypeIs
+
+def _(
+    a: TypeGuard[str],
+    b: TypeIs[str | int],
+    c: TypeGuard[Intersection[complex, Not[int], Not[float]]],
+    d: TypeIs[tuple[TypeOf[bytes]]],
+):
+    reveal_type(a)  # revealed: TypeGuard[str]
+    reveal_type(b)  # revealed: TypeIs[str | int]
+    reveal_type(c)  # revealed: TypeGuard[complex & ~int & ~float]
+    reveal_type(d)  # revealed: TypeIs[tuple[Literal[bytes]]]
+
+def f(a) -> TypeGuard[str]: ...
+def g(a) -> TypeIs[str]: ...
+
+def _(a: object):
+    reveal_type(f(a))  # revealed: TypeGuard[a, str]
+    reveal_type(g(a))  # revealed: TypeIs[a, str]
+```
+
+## Parameters
+
+A user-defined type guard must accept at least one positional argument, (in addition to `self`/`cls`
+for non-static methods).
+
+```py
+from typing_extensions import TypeGuard, TypeIs
+
+# error: [invalid-type-guard-definition]
+def _() -> TypeGuard[str]: ...
+
+# error: [invalid-type-guard-definition]
+def _(**kwargs) -> TypeIs[str]: ...
+
+class _:
```

Comment:
I'm not sure if it's possible or reasonable here but I found many very specific mdtests much more helpful when debugging regression than a few tests that have many unrelated assertions. 

There are also fewer trade-offs when using smaller tests compared to Ruff because it doesn't require you to create a new file. A single header is enough to create an isolated test.